pkgs/applications/science/math/caffe/default.nix

```diff
@@ -73,9 +73,17 @@ stdenv.mkDerivation rec {
                 ++ lib.optionals stdenv.isDarwin [ Accelerate CoreGraphics CoreVideo ]
                 ;
 
-  propagatedBuildInputs = lib.optional pythonSupport python.pkgs.protobuf;
-
-  outputs = [ "bin" "out"];
+  propagatedBuildInputs = lib.optionals pythonSupport (
+    # requirements.txt
+    let pp = python.pkgs; in ([
```

Comment:
Why not `with python.pkgs [ numpy scipy ... ]`?

Reply:
I did try that initially, but `protobuf` seemed to clash with the above import (didn't install).
Perhaps I messed up something else, but this was my workaround

Reply:
Please cherry-pick https://github.com/matthiasbeyer/nixpkgs/commit/879db97fb5a85f2d2dcb6c28231f9ba382cc2e9f

With this patch, `caffe` still builds for me.

Reply:
The build works regardless of the clash, simply start up python and `import caffe`.
```
Python 3.7.3 (default, Mar 25 2019, 20:59:09)
[GCC 7.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import caffe
Failed to include caffe_pb2, things might go wrong!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/30n7msvdrqhj4ab9151rp402j50vdsck-caffe-1.0/lib/python3.7/site-packages/caffe/__init__.py", line 4, in <module>
    from .proto.caffe_pb2 import TRAIN, TEST
  File "/nix/store/30n7msvdrqhj4ab9151rp402j50vdsck-caffe-1.0/lib/python3.7/site-packages/caffe/proto/caffe_pb2.py", line 7, in <module>
    from google.protobuf.internal import enum_type_wrapper
ModuleNotFoundError: No module named 'google'
```

Reply:
How do you even import this, the package seems to provide only binaries, no libraries... or am I completely missing something here?

Reply:
```
$ nix-shell -I nixpkgs=<nkpkgs-root-dir> -p python3Packages.caffe
$ python
```

The python libraries specified get added to the `PYTHONPATH` env var (which python uses for imports).

If you echo out `PYTHONPATH` and check for caffe, you should be able to find the copied sources there (at least on my branch).

Reply:
in nixpkgs tree:

```
$ nix-shell -p python3Packages.caffe -p python3
$ which python
/nix/store/gbwx5sc4w0i0ipnmsa0nsk88svs983vl-python3-3.7.2/bin/python
$ python
import caffe
```

works as expected for me (including my patch, on a centos VM).

Reply:
You don't need to specify python3 as an extra package, python3xxx internally requires it.
I'm on nixos 19.03. Try adding the pure arg `--pure to nix-shell`. Your patch is not working for me.

Reply:
```
$ nix-shell -p python3Packages.caffe --pure --run "python"
```

And then importing `caffe` works for me, I'm on centos7 on my testing machine. Fun things... maybe someone can explain why this works for me...

Reply:
If you're running centos 7, do you have python bindings `protobuf` elsewhere in the env? `rpm -q protobuf-python` perhaps? Maybe try importing `google` and `print(google.__file__)`. Aside from that, I have no idea

Reply:
Nope, none of that. Also, `--pure` would prevent leaking stuff into the environment, right?

Reply:
> If this flag is specified, the environment is almost entirely cleared before the interactive shell is started, so you get an environment that more closely corresponds to the “real” Nix build. A few variables, in particular HOME, USER and DISPLAY, are retained. Note that ~/.bashrc and (depending on your Bash installation) /etc/bashrc are still sourced, so any variables set there will affect the interactive shell.

AKA, if `PYTHONPATH` is leaking in somehow, that might explain why.
I'm off to sleep now friend. G'night